common/hal/hal_vw_gpio.c

```diff
@@ -0,0 +1,142 @@
+{
+#ifdef CONFIG_ESPI_ASPEED
+	// Check if the vw gpio is master input and bic output pin
+	uint32_t reg_value = sys_read32(AST_ESPI_BASE + AST_ESPI_GPIO_DIR);
+	if (!GETBIT(reg_value, number))
+		return false;
+
+	uint8_t i;
+	for (i = 0; i < vw_gpio_size; i++) {
+		if ((vw_gpio_cfg[i].number == number) && (vw_gpio_cfg[i].is_enabled) &&
+		    (vw_gpio_cfg[i].direction == VW_GPIO_OUTPUT)) {
+			uint32_t reg_value = sys_read32(AST_ESPI_BASE + AST_ESPI_GPIO_VAL);
+			if (value == VW_GPIO_HIGH)
+				reg_value = SETBIT(reg_value, number);
+			else
+				reg_value = CLEARBIT(reg_value, number);
+			sys_write32(reg_value, AST_ESPI_BASE + AST_ESPI_GPIO_VAL);
+			return true;
+		}
+	}
+	return false;
+#else
+	return false;
+#endif
+}
+
+static void ast_vw_gpio_scan(void)
+{
+	uint8_t i;
+	for (i = 0; i < vw_gpio_size; i++) {
+		if (!vw_gpio_cfg[i].is_enabled)
+			continue;
+
+		if (vw_gpio_cfg[i].direction == VW_GPIO_INPUT) {
+			/* Check if controller direction setting
+			 * ESPI9C: GPIO direction of virtual wire channel
+			 */
+			uint32_t reg_value = sys_read32(AST_ESPI_BASE + AST_ESPI_GPIO_DIR);
+			if (GETBIT(reg_value, vw_gpio_cfg[i].number) != VW_GPIO_INPUT) {
+				LOG_ERR("the vgpio setting at controller side is incorrect.");
+				continue;
+			}
+
+			/* Get virtual gpio value
+			 * ESPI09C: GPIO through virtual wire channel
+			 */
+			reg_value = sys_read32(AST_ESPI_BASE + AST_ESPI_GPIO_VAL);
+			uint8_t gpio_value = GETBIT(reg_value, vw_gpio_cfg[i].number) ?
+						     VW_GPIO_HIGH :
+						     VW_GPIO_LOW;
+			if (gpio_value != vw_gpio_cfg[i].value) {
+				vw_gpio_cfg[i].value = gpio_value;
+				if (vw_gpio_cfg[i].int_cb)
+					vw_gpio_cfg[i].int_cb(gpio_value);
+			}
+		}
+	}
+}
+
+/* Handler for virtual GPIO from eSPI virtual wire channel */
+static void vw_handler(const struct device *dev, struct espi_callback *cb, struct espi_event event)
+{
+#ifdef CONFIG_ESPI_ASPEED
+	if (event.evt_type == ESPI_BUS_EVENT_VWIRE_RECEIVED) {
+		/* Check if the virtual wire GPIO interrupt is happened
+		 * ESPI008: Interrupt Status
+		 *   bit[9] - Virtual Wire GPIO Event
+		 */
+		if (!GETBIT(event.evt_data, 9))
+			return;
+
+		ast_vw_gpio_scan();
+	}
+#endif
+}
+
+bool vw_gpio_init(vw_gpio *config, uint8_t size)
+{
+	CHECK_NULL_ARG_WITH_RETURN(config, false);
+
+	espi_dev = device_get_binding("ESPI");
+	if (!espi_dev) {
+		LOG_ERR("failed to get espi device");
+		return false;
+	}
+
+	vw_gpio_cfg = (vw_gpio *)malloc(size * sizeof(vw_gpio));
+	if (!vw_gpio_cfg) {
+		LOG_ERR("failed to allocate memory");
+		return false;
+	}
+	memcpy(vw_gpio_cfg, config, size * sizeof(vw_gpio));
+	vw_gpio_size = size;
+
+	espi_init_callback(&vw_cb, vw_handler, ESPI_BUS_EVENT_VWIRE_RECEIVED);
+	if (espi_add_callback(espi_dev, &vw_cb) < 0) {
+		LOG_ERR("failed to add espi callback function");
+		SAFE_FREE(vw_gpio_cfg);
+		return false;
+	}
+
+	// Initialize vGPIO status during BIC bootup
+#ifdef CONFIG_ESPI_ASPEED
+	ast_vw_gpio_scan();
+#endif
+
```

Comment:
Is this file only called once?  Otherwise, we have the risk of a memory leak on vw_gpio_cfg.

Reply:
Thanks for your comment.
This file is only called once but the eSPI callback function "vw_cb" is always existed in background.
The function "vw_handler" is fired by zephyr eSPI driver if there is any eSPI virtual wire interrupt.
And the vw_gpio_cfg is used in vw_handler function.
So, the vw_gpio_cfg cannot be freed in vw_gpio_init function.
Additionally, we reserved for the varity chips in this file such Aspeed, TI, Nuvoton BIC chips using CONFIG_ESPI_ASPEED.
Now, we only support Aspeed chip. The vw_gpio_cfg can be freed if the chip is not Aspeed BIC.
==============================================
	// Initialize vGPIO status during BIC bootup
#ifdef CONFIG_ESPI_ASPEED
	ast_vw_gpio_scan();
	return true;
#endif

	SAFE_FREE(vw_gpio_cfg);
	vw_gpio_size = 0;
	return false;
==============================================

BTW, we has contributed the zephyr patch to Aspeed and Aspeed will include the eSPI callback mechanism in next Aspeed SDK release.

The detail eSPI callback mechanism is in the zephyr document.
https://docs.zephyrproject.org/apidoc/latest/group__espi__interface.html#ga8d88e4e3893d610713195e5352ec2565